### PR TITLE
[#12] handle memory leaks

### LINKED ISSUE #12: Handle memory leaks
### Details:

When we add event listeners to our elements, we bind them to the DOM and we never remove them in the current code base.
This leads to the uncontrolled memory leaking. Check out the code bellow:

```javascript
// two buttons
document.body.insertAdjacentHTML(
  'beforeend', 
  `<button id="gonna_be_removed_at_once"></button>
   <button id="action">Click Me</button>`
);

// the one we gonna get rid of
const removed= document.body.querySelector('#gonna_be_removed_at_once');
// adding listener to this to see the problem
removed.addEventListener('click', () => console.log('Hello from nowhere!'));
// see, there is no btn with the id "gonna_be_removed_at_once" on the page
removed.remove(); 

// the action button we can click to emit event
const actionBtn= document.body.querySelector('#action');
// every time I click the action button it imitates click on the "removed" button
actionBtn.addEventListener('click', () => removed.click());

// click -> console: Hello from nowhere!
// click -> console: Hello from nowhere!
// click -> console: Hello from nowhere!
// click -> console: Hello from nowhere!
```

You can play with it on this [codepen](https://codepen.io/Vladlen-Volkov/pen/bGQeMVN). 
As you can see there is no button with id "gonna_be_removed_at_once" on the page, bc we've completely removed this, right? But it's still in the RAM (random access memory / "оперативная память") and moreover it pulls everything from its closure... For the modern PCs it's not a big deal, however it rapidly becomes a problem when we don't pay enough attention to this. 

The same problem occurs when you remove `todoList` and `listItem` in our app.

To prevent this problem we just need to clean the element from the listeners:

```javascript
const removed= document.body.querySelector('#gonna_be_removed_at_once');
const greeting = () => console.log('Hello from nowhere!'); // 1.extract handlers to the variables
removed.addEventListener('click', greeting ); 

removed.remove(); 
removed.removeEventListener('click', greeting );  // 2. remove listeners when you remove the element itself
```

You might find this [section](https://learn.javascript.ru/introduction-browser-events#addeventlistener) about removing listeners helpful in solving the problem.

### PR DESCRIPTION
**Pull request by trofimovama**
closes #12

### CODE REVIEW

File: BestPractice.md [hunk 1]
```diff
@@ -1,3 +1,4 @@
 1. Функции всегда с маленькой буквы, если это не компоненты и не классы;
 2. Если оператор присваивания вызывается только 1 раз, то это всегда "const";
-3. Если сравниваешь строки, то приводим к одному единому регистру (`.toLowerCase()`);
\ No newline at end of file
+3. Если сравниваешь строки, то приводим к одному единому регистру (`.toLowerCase()`);
+4. Event Listeners должны быть удалены, где необходимо, для того, чтобы избежать утечки памяти;
```
Comment (vladlen-volkov) [inline BestPractice.md:4]: 👍

File: script.js [hunk 2]
```diff
@@ -134,23 +134,27 @@ const renderTodoList = (todoListState, onDestroy) => {
...
+  }
 
-  listButton.addEventListener('click', () => createNewTodoItem());
+  inputField.addEventListener('keydown', inputActions);
+
+  listButton.addEventListener('click', createNewTodoItem);
 
   deleteCardIcon.addEventListener('click', () => {
     cardContainer.remove();
+    inputField.removeEventListener('keydown', inputActions);
```
Comment (vladlen-volkov) [inline script.js:154]: Missed cleaning up from click listener added on the L:150 `listButton.addEventListener('click', createNewTodoItem);`

File: script.js [hunk 3]
```diff
@@ -134,23 +134,27 @@ const renderTodoList = (todoListState, onDestroy) => {
     notifyStateUpdated();
   };
 
-  inputField.addEventListener('keydown', (e) => {
+  const escapeEmpty = () => inputField.value = "";
```
Comment (vladlen-volkov) [inline script.js:137]: From my perspective this function should be called `cleanField`. 
It's a minor,  I leave it up to you

File: script.js [hunk 4]
```diff
@@ -114,7 +114,7 @@ const renderTodoList = (todoListState, onDestroy) => {
 
   const createNewTodoItem = () => {
     if (!inputField.value) {
-      return;
+      inputField.removeEventListener('keydown',  createNewTodoItem());
```
Comment (vladlen-volkov) [inline script.js:117]: Let's rollback the line as we decided on the call

**Review by vladlen-volkov — CHANGES_REQUESTED** [review]
Almost there, come on 💪

File: script.js (hunk 3 above)
Comment (trofimovama) [reply script.js]: Yes, you are right. Renamed)

File: script.js (hunk 4 above)
Comment (trofimovama) [reply script.js]: OMG! Missed here) Fixed forever!)

**Review by vladlen-volkov — APPROVED** [review]

**Pull request merged**